File: Crash it/Shuttle/ShuttleEnemy.swift
//
//  Enemy.swift
//  Crash it
//
//  Created by Olivier Picard on 13/03/2018.
//  Copyright © 2018 Olivier Picard. All rights reserved.
//

import Foundation
import SpriteKit

class ShuttleEnemy: Shuttle {
    let target: ShuttlePlayer
    var deltaTime_check_targetPosition: TimeInterval
    private var previous_time_updatedDirection: TimeInterval
    
    
    init(image: UIImage, color: UIColor, stats: Shuttle.Stats, target: ShuttlePlayer) {
        self.target = target
        self.deltaTime_check_targetPosition = 2
        self.previous_time_updatedDirection = 0
        super.init(image: image, color: color, stats: stats)
        super.speed_factor = 2.0
        super.lifeBar.position = CGPoint(x: self.position.x,
                                         y: self.position.y + self.size.height/2 + super.lifeBar.size.height/2 + 8)
        
    }
    
    required init?(coder aDecoder: NSCoder) {
        self.target = ShuttlePlayer(coder: aDecoder)!
        self.previous_time_updatedDirection = 0
        self.deltaTime_check_targetPosition = 2
        super.init(coder: aDecoder)
    }
    
    override func update(_ currentTime: TimeInterval) {
        super.update(currentTime)
        if self.stats.shoot_stats.canShoot(currentTime) {
            shoot(direction: self.direction, rotation: self.zRotation)
        }
        if(currentTime - self.previous_time_updatedDirection >= self.deltaTime_check_targetPosition
            && self.position.y - target.position.y > self.target.position.y + self.scene!.size.height/15) {
            
            self.previous_time_updatedDirection = currentTime
            updateDirectionToTarget()
        }
    }
    
    func updateDirectionToTarget() {
        // Soit un triangle rectangle en A, B est self et C la target
        let len_BA = self.position.y - target.position.y
        let len_CA = target.position.x - self.position.x
        
        let angleRadian = atan(len_CA/len_BA)
        self.zRotation = angleRadian
        self.direction = CGVector(dx: sin(angleRadian), dy: -cos(angleRadian))
        
    }

    override func inCollisionWith(item: Collisionable) {
        var increaseScore = 0
        if let laser = item as? LaserShot {
            if type(of: laser.shooter) == ShuttleEnemy.self { return }
            self.lifeBar.value -= laser.shooter.stats.attack
            if self.lifeBar.value == 0 { increaseScore += 1 }
        } else if let _ = item as? ShuttlePlayer {
            self.lifeBar.value = 0
            increaseScore += 1
        }
        
        if(increaseScore > 0) {
            let gameScene = self.scene! as! GameScene
            gameScene.increaseScore(1)
            gameScene.removeChildren(in: [self])
        }
    }
    
}

File: Android/app/src/main/java/com/olivierpicard/crachit/Shuttle/ShuttleEnemy.java
package com.olivierpicard.crachit.Shuttle;

import com.olivierpicard.crachit.AnimatedItem.Explosion;
import com.olivierpicard.crachit.GameScene;
import com.olivierpicard.crachit.Graphics.GPoint;
import com.olivierpicard.crachit.Graphics.GVector;
import com.olivierpicard.crachit.ICollisionable;
import com.olivierpicard.crachit.LaserShot;
import com.olivierpicard.crachit.Tools;

/**
 * Created by olivierpicard on 06/04/2018.
 */

public class ShuttleEnemy extends Shuttle {
    final ShuttlePlayer target;
    double deltaTime_check_targetPosition;
    private double previous_time_updatedDirection;


    ShuttleEnemy(int bitmapRessourceID, int color, Stats stats, ShuttlePlayer target) {
        super(bitmapRessourceID, color, stats);
        this.target = target;
        this.deltaTime_check_targetPosition = 2;
        this.previous_time_updatedDirection = 0;
        this.speed_factor = 2f;
        this.direction = GVector.down();
        this.lifeBar.setPosition(new GPoint(this.getPosition().x,
                this.getPosition().y - this.getSize().height/2
                        - this.lifeBar.getSize().height/2 - 8));
    }


    @Override
    public void update(long currentTime) {
        super.update(currentTime);
        if(this.stats.shoot_stats.canShoot(currentTime))
            shoot(this.direction, this.getZRotation());

        if(this.getScene() != null
                && currentTime - this.previous_time_updatedDirection >= this.deltaTime_check_targetPosition
                && target.getPosition().y - this.getPosition().y > this.target.getPosition().y - Tools.screenMetrics.heightPixels/15)
        {
            this.previous_time_updatedDirection = currentTime;
            updateDirectionToTarget();
        }
    }


    public void updateDirectionToTarget() {
        final float len_BA = this.getPosition().y - target.getPosition().y;
        final float len_CA = target.getPosition().x - this.getPosition().x;
        final float angle = (float)Math.atan(len_CA/len_BA);
        setZRotation((float)(angle*180/Math.PI));
        this.direction = new GVector(-Math.sin(angle), Math.cos(angle));
    }


    @Override
    public void inCollisionWith(ICollisionable item) {
        int increaseScore = 0;
        if(item instanceof LaserShot) {
            final LaserShot laser = (LaserShot)item;
            if(laser.shooter instanceof ShuttleEnemy) return;
            this.lifeBar.setValue(this.lifeBar.getValue() - laser.shooter.stats.attack);
//              TODO : increaseScore
//            if(this.lifeBar.getValue() == 0) increaseScore += 1;
        } else if(item instanceof ShuttlePlayer) {
            this.lifeBar.setValue(0);
            increaseScore += 1;
        }

        if(increaseScore > 0) {
            final GameScene scene = (GameScene)getScene();
//            TODO : Remplir la class Explosion
//            scene.addChild(new Explosion(this.getPosition()));
            this.getScene().removeChild(this);
        }
    }


}
